### PR TITLE
fix: merge collection element schemas with differing key sets into array

### LINKED ISSUE #527: Collections of maps with different key sets are converted to indexed structs instead of arrays
# Collections of maps with different key sets are converted to indexed structs instead of arrays

## Problem

When `CompactValueConverter` encounters a collection containing maps with different key sets, it converts the collection into an indexed struct (`{e0: {...}, e1: {...}}`) instead of preserving it as an array.

This affects any Cypher source query that uses `collect()` over maps where optional relationships or conditional logic produces elements with varying structures.

### Case 1: Key absent in some elements

```kotlin
val coll = listOf(
  mapOf("name" to "john", "age" to 21),
  mapOf("name" to "jane", "age" to 25, "address" to mapOf("city" to "london", "zip" to 10001))
)
val schema = converter.schema(coll, false)
```

**Expected**: `schema.type()` is `ARRAY` with a merged element schema containing `name`, `age`, and optional `address`.

**Actual**: `schema.type()` is `STRUCT` with indexed fields `e0`, `e1`.

### Case 2: Key present but value is null

```kotlin
val coll = listOf(
  mapOf("name" to "john", "age" to 21, "address" to null),
  mapOf("name" to "jane", "age" to 25, "address" to mapOf("city" to "london", "zip" to 10001))
)
val schema = converter.schema(coll, false)
```

**Expected**: Same as Case 1 — `ARRAY` with optional `address`.

**Actual**: Fails at value conversion with `DataException: Invalid value: null used for required field: "address", schema type: STRUCT`.

Case 2 is the more common scenario in practice, as it arises from Cypher patterns like:

```
collect({ id: node.id, detail: CASE WHEN rel IS null THEN null ELSE {x: rel.x} END })
```

The map literal always includes the `detail` key, but with a null value when the optional relationship does not exist.

## Root cause

In `CompactValueConverter.schema()`, two separate mechanisms cause schema divergence:

### Case 1: Key absent — `notNullOrEmpty()` excludes null entries from key set

The Map branch uses `notNullOrEmpty()` to filter entries before schema inference. When a key is entirely absent from one element but present in another, the inferred schemas have different key sets. The Collection branch then compares schemas via `toSet().size`, finds multiple unique schemas, and falls back to the indexed struct `{e0, e1, ...}`.

### Case 2: Key present with null value — `schema(null)` returns OPTIONAL_STRING

When a key IS present but its value is null, the Map branch's `else` case includes it in the schema by calling `schema(null, ...)`, which returns `SimpleTypes.NULL.schema(true)` — an `OPTIONAL_STRING_SCHEMA`. If the same key in another element has a non-null structured value (producing a `STRUCT` schema), the two schemas have the same key set but **incompatible types** (`STRING` vs `STRUCT`) for that field.

This is worse than Case 1: schema merging cannot resolve the type conflict, but the schemas also cannot be treated as identical. In older versions without merging, this falls back to `{e0, e1}`. With naive merging that does not account for null schemas, it either falls back or (if a different code path infers the schema from a single element) produces a schema with `address` as a **required** `STRUCT`, causing `DataException` at value conversion time when a null value is encountered.

### Both cases

The `force-maps-as-struct` setting does not help in either case, as it only affects maps with homogeneous value types, not collections with heterogeneous element schemas.

### 1. Schema merging for collections with differing key sets

When the Collection branch encounters multiple distinct STRUCT schemas, attempt to **merge** them before falling back to the indexed struct format:

1. Check that all schemas are of type `STRUCT`
2. Compute the union of all field names across schemas
3. For each field, resolve the type from schemas that have it
4. Fields present in only a subset of schemas are marked as **optional**
5. If all field types are compatible (same type, or recursively mergeable structs), return a merged schema and use it as the array element schema
6. If types conflict (e.g., same field is `STRING` in one and `INT64` in another), fall back to the existing `{e0, e1, ...}` behavior

### 2. Null schema compatibility

During schema merging, the null schema (`SimpleTypes.NULL.schema(true)` = `OPTIONAL_STRING_SCHEMA`) is treated as compatible with any other type. When a field has a null schema in some elements and a concrete type in others:

- The concrete type is used as the resolved schema
- The field is marked as **optional** (since null indicates the value may be absent)

If all elements have null schema for a field, the null schema itself is used (as optional).

### Backward compatibility

Both changes are backward-compatible:

- Incompatible schemas (e.g., same field is `INT64` in one element and `STRUCT` in another) still fall back to the existing `{e0, e1, ...}` behavior
- No changes to `value()` are needed — the existing Map-to-Struct conversion iterates schema fields and looks up values by key. Missing keys and null values return `null`, which is valid for optional fields.

## Workaround

Without this fix, users must ensure all collection elements have identical key sets and types by using `coalesce` to provide typed default values:

```cypher
collect({
  id: node.id,
  detail: { x: coalesce(rel.x, 0), y: coalesce(rel.y, '') }
})
```

This preserves the array format but introduces empty/zero defaults for null values, which may cause issues in downstream systems (e.g., Elasticsearch date field parsing, enum validation).

### PR DESCRIPTION
**Pull request by ogawa-takeshi**
## Summary

Fixes #527

When `CompactValueConverter` encounters a collection of maps with different key sets (common with `collect()` over optional relationships), it falls back to indexed struct format `{e0, e1, ...}` instead of preserving the collection as an array. This PR adds schema merging that computes the union of fields and marks subset-only fields as optional.

## Changes

- Add `mergeStructSchemas()` to `CompactValueConverter` — merges multiple STRUCT schemas by taking the union of fields, marking fields absent from some schemas as optional
- Handle null schema compatibility — `schema(null)` returns `OPTIONAL_STRING`, which is now treated as compatible with any concrete type during merging
- Add `makeOptional()` helper for wrapping non-optional schemas
- Add 12 test cases covering: key-absent pattern, null-value pattern, recursive merge, order independence, 3+ elements, all-null fields, optional parameter, and fallback scenarios

### Case 1: Key absent in some elements
`notNullOrEmpty()` filter excludes null map entries from schema inference, producing different key sets across elements. The Collection branch sees multiple unique schemas and falls back to `{e0, e1}`.

### Case 2: Key present with null value
`schema(null)` returns `OPTIONAL_STRING_SCHEMA`. When the same key has a `STRUCT` in other elements, the type mismatch (`STRING` vs `STRUCT`) prevents merging and causes `DataException` at value conversion.

## Backward compatibility

- Incompatible field types (e.g., `INT64` vs `STRUCT` for the same field) still fall back to the existing `{e0, e1, ...}` behavior
- No changes to `value()` — the existing Map-to-Struct conversion handles missing keys and null values naturally for optional fields

### CODE REVIEW

**Comment by venikkin** [comment]
Hi @ogawa-takeshi, thank you for signing the CLA. It appears that you might have forgotten to attach `neo4j-cla.pdf`. Would you mind re-sending the email with the PDF attached? Once we receive it, someone from the team will review your PR and get back to you.

**Comment by ogawa-takeshi** [comment]
@venikkin 
Thank you for letting me know. I have re-sent the email with neo4j-cla.pdf attached.

**Comment by ogawa-takeshi** [comment]
@venikkin Hi, just following up — I re-sent the CLA PDF on March 6th. Could you confirm if it was received? I'd love to get this PR reviewed when possible. Thanks!

**Comment by fbiville** [comment]
Hello, 

Thanks for your patience and contribution.

I can confirm we have received your CLA and have 
reviewed your pull request.

Regarding what you describe as "Case 1", the new 
schema logic seems to be skipping some keys from 
the data. 
If you add `val actualValue = converter.value(schema, coll)` 
after `val schema = converter.schema(coll, false)`
in the tests in `CompactValueConverterTest`, you 
will see the tests start to fail.

Moreover, we initially intended to avoid scanning
all data for schema inference as this is potentially
costly.
We might only go ahead with this strategy for the 
next major version of the connector, as this is 
quite a big change.

For now, we would be quite interested in getting
"Case 2" fixed.

Would you willing to refocus this PR on "Case 2" 
only and optionally extract the handling of 
"Case 1" in a separate PR?

**Comment by ogawa-takeshi** [comment]
Hi @fbiville, thanks for the detailed review and for confirming receipt of the CLA.

I've split the work per your request:

- **This PR (#528) is now Case 2 only** — same key, null in some elements vs. concrete (typically STRUCT) in others. Force-pushed `fix/collection-schema-merge` to a single commit on top of latest `main`. The merge is deliberately conservative: it requires all element Schemas to be STRUCT with **identical field name sets**, and only resolves a NULL-typed schema against a single concrete type per field. Anything else still falls back to the existing indexed-struct `{e0, e1, ...}` representation, so `value()` stays consistent with whatever schema is produced. Every new test calls `converter.value(schema, coll)` and asserts on the result.

- **Case 1 is now #573** — different key sets across elements. That rewrite drops the buggy `mergeMapAndStructSchemas(Set<Schema>)` approach (which is what was silently dropping keys, exactly as you described) and instead merges from the raw `Map` values with `forceMapsAsStruct=true`, so MAP-shaped and STRUCT-shaped element representations are reconciled before any keys can be lost. The PR description includes a test specifically reproducing the regression you flagged.

On the broader concern about scanning all data for schema inference: I tried to keep both PRs minimal — the merge runs only when `schema()` was already iterating the Collection's elements (the existing fallback path). #528 in particular only inspects the per-element Schemas it had already computed; #573 does walk the raw values one extra level. If even that is more than you'd like before the next major, please feel free to defer #573 — #528 alone unblocks the most common variant of #527 in our use case.

Happy to revise either PR based on further feedback. Thanks again!

File: common/src/main/kotlin/org/neo4j/connectors/kafka/data/converter/CompactValueConverter.kt [hunk 1]
```diff
@@ -304,4 +318,63 @@ class CompactValueConverter : ValueConverter {
...
+      val resolvedSchema =
+          when {
+            nonNullFieldSchemas.isEmpty() -> nullSchema
+            nonNullFieldSchemas.size == 1 -> nonNullFieldSchemas.first()
+            nonNullFieldSchemas.all { it.type() == Schema.Type.STRUCT } ->
+                mergeNullableStructSchemas(nonNullFieldSchemas) ?: return null
+            else -> return null
+          }
+
+      builder.field(fieldName, if (hasNullSchema) makeOptional(resolvedSchema) else resolvedSchema)
```
Comment (fbiville) [inline common/src/main/kotlin/org/neo4j/connectors/kafka/data/converter/CompactValueConverter.kt:359]: suggestion: let's always make the schema optional, in case a future message misses some fields

File: common/src/test/kotlin/org/neo4j/connectors/kafka/data/converter/CompactValueConverterTest.kt [hunk 2]
```diff
@@ -703,4 +704,143 @@ class DynamicTypesCompactTest {
...
+    val schema = converter.schema(coll, false)
+    val actualValue = converter.value(schema, coll)
+
+    schema.type() shouldBe Schema.Type.STRUCT
+    schema.fields().map { it.name() } shouldBe listOf("e0", "e1")
+    actualValue.shouldBeInstanceOf<Struct>()
+  }
+
+  @Test
+  fun `collections with conflicting non-null field types should fall back to indexed struct schema`() {
```
Comment (fbiville) [inline common/src/test/kotlin/org/neo4j/connectors/kafka/data/converter/CompactValueConverterTest.kt:792]: suggestion: add a similar test, with nested values (same field names, different types)

**Review by fbiville — COMMENTED** [review]
Thanks, the PR looks good. Just a couple of small suggestions.

File: common/src/main/kotlin/org/neo4j/connectors/kafka/data/converter/CompactValueConverter.kt (hunk 1 above)
Comment (ogawa-takeshi) [reply common/src/main/kotlin/org/neo4j/connectors/kafka/data/converter/CompactValueConverter.kt]: Done in fdf77132 — every field of the merged struct is now marked optional regardless of whether a null was observed, and the existing tests were updated accordingly.

File: common/src/test/kotlin/org/neo4j/connectors/kafka/data/converter/CompactValueConverterTest.kt (hunk 2 above)
Comment (ogawa-takeshi) [reply common/src/test/kotlin/org/neo4j/connectors/kafka/data/converter/CompactValueConverterTest.kt:793]: Added in fdf77132: `nested values with same field names but conflicting types should fall back to indexed struct schema`, where both `addr` entries are structs with the same field names but conflicting `city`/`zip` types, so the recursive merge declines and `value()` still succeeds.

**Review by venikkin — APPROVED** [review]
